### PR TITLE
feat(mosaicod)!: Improve existing API-KEY tests coverage

### LINKED ISSUE #426: Improve existing API-KEY tests Coverage in mosaicod
While some API-KEY backend tests already exist, they do not cover important edge cases.

Current tests do not verify that API keys are actually enforced and used correctly.
The following cases should be added:

- [x] Valid READ API key: request of read succeeds as expected.
- [x] Valid WRITE API key: request of write succeeds as expected.
- [x] Invalid WRITE API key: request of write fails as expected.
- [x] Valid DELETE API key: request of delete succeeds as expected.
- [x] Invalid DELETE API key: request of delete fails as expected.
- [x] Valid MANAGE API key: request of manage succeeds as expected.
- [x]  Invalid MANAGE API key: request of manage fails as expected.

### PR DESCRIPTION
**Pull request by LuigiTer**
* Added tests for API key Read/Write/Delete/Manage.
* Fixed expired API key usage.

BREAKING CHANGES

* Fixed inconsistent returned error code.
* Fixed error return for sequence/topic/session delete.
 
Closes  https://github.com/mosaico-labs/mosaico/issues/426 https://github.com/mosaico-labs/mosaico/issues/458 https://github.com/mosaico-labs/mosaico/issues/459 https://github.com/mosaico-labs/mosaico/issues/460

### CODE REVIEW

File: mosaicod/crates/mosaicod-core/src/types/auth.rs [hunk 1]
```diff
@@ -36,7 +36,7 @@ pub enum ApiKeyError {
 impl PublicError for ApiKeyError {
     fn error(&self) -> Error {
         match self {
-            Self::MissingPermissions => Error::unauthorized(),
+            Self::MissingPermissions => Error::unauthorized("Missing permissions".to_string()),
```
Comment (DevMattG) [inline mosaicod/crates/mosaicod-core/src/types/auth.rs]:
```suggestion
            Self::MissingPermissions => Error::unauthorized("missing permissions".to_string()),
```

File: mosaicod/crates/mosaicod-facade/src/topic.rs [hunk 2]
```diff
@@ -126,7 +126,9 @@ pub async fn try_create(
     let seq_rec = db::sequence_find_by_locator(&mut tx, session_handle.sequence_locator()).await?;
 
     if !locator.is_sub_locator(session_handle.sequence_locator()) {
-        Err(core::Error::unauthorized())?;
+        Err(core::Error::unauthorized(
+            "Provided topic locator and session do not share the same sequence".to_string(),
```
Comment (DevMattG) [inline mosaicod/crates/mosaicod-facade/src/topic.rs]:
```suggestion
            "provided topic locator and session do not share the same sequence".to_string(),
```

File: mosaicod/crates/mosaicod-marshal/src/actions/core.rs [hunk 3]
```diff
@@ -108,6 +108,33 @@ pub enum ActionRequest {
     Version(requests::Empty),
 }
 
+impl std::fmt::Display for ActionRequest {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::SequenceCreate(_) => write!(f, "SequenceCreate()"),
```
Comment (DevMattG) [inline mosaicod/crates/mosaicod-marshal/src/actions/core.rs]: Remove parenthesis after action name, e.g. "SequenceCreate"

File: mosaicod/crates/mosaicod-server/src/endpoint/do_action.rs [hunk 4]
```diff
@@ -20,7 +20,8 @@ pub async fn do_action(
     perm: &Permission,
 ) -> Result<ActionResponse> {
     if !has_permissions(&action, perm) {
-        Err(core::Error::unauthorized())?;
+        let err_msg = format!("{} has not enough permission to be executed.", action);
```
Comment (DevMattG) [inline mosaicod/crates/mosaicod-server/src/endpoint/do_action.rs]:
```suggestion
        let err_msg = format!("provided API key has not enough permissions to execute {} action.", action);
```

File: mosaicod/crates/mosaicod-server/src/endpoint/do_put.rs [hunk 5]
```diff
@@ -97,7 +97,9 @@ async fn do_put_topic_data(
         .map_err(|_| core::Error::bad_uuid(uuid_str.clone()))?;
 
     if received_uuid != topic_uuid {
-        Err(core::Error::unauthorized())?
+        Err(core::Error::unauthorized(
+            "Received uuid does not match the topic uuid.".to_string(),
```
Comment (DevMattG) [inline mosaicod/crates/mosaicod-server/src/endpoint/do_put.rs]:
```suggestion
            "received uuid does not match the topic uuid.".to_string(),
```

File: mosaicod/crates/mosaicod-server/src/flight.rs [hunk 6]
```diff
@@ -227,7 +227,9 @@ impl MosaicodFlight {
         let auth_ctx = auth_context(&request)?;
 
         if !auth_ctx.permissions().can_read() {
-            Err(core::Error::unauthorized())?;
+            Err(core::Error::unauthorized(
+                "Provided API key does not have READ permissions.".to_string(),
```
Comment (DevMattG) [inline mosaicod/crates/mosaicod-server/src/flight.rs]:
```suggestion
                "provided API key does not have READ permissions.".to_string(),
```

File: mosaicod/crates/mosaicod-server/src/flight.rs [hunk 7]
```diff
@@ -244,7 +246,9 @@ impl MosaicodFlight {
...
```
Comment (DevMattG) [inline mosaicod/crates/mosaicod-server/src/flight.rs]:
```suggestion
                "provided API key does not have READ permissions.".to_string(),
```

File: mosaicod/crates/mosaicod-server/src/flight.rs [hunk 8]
```diff
@@ -260,7 +264,9 @@ impl MosaicodFlight {
     async fn impl_do_get(&self, request: Request<Ticket>) -> Result<Response<DoGetStream>> {
         let auth_ctx = auth_context(&request)?;
         if !auth_ctx.permissions().can_read() {
-            Err(core::Error::unauthorized())?;
+            Err(core::Error::unauthorized(
+                "Provided API key does not have READ permissions.".to_string(),
```
Comment (DevMattG) [inline mosaicod/crates/mosaicod-server/src/flight.rs]:
```suggestion
                "provided API key does not have READ permissions.".to_string(),
```

File: mosaicod/crates/mosaicod-server/src/flight.rs [hunk 9]
```diff
@@ -281,7 +287,9 @@ impl MosaicodFlight {
     ) -> Result<Response<DoPutStream>> {
         let auth_ctx = auth_context(&request)?;
         if !auth_ctx.permissions().can_write() {
-            Err(core::Error::unauthorized())?;
+            Err(core::Error::unauthorized(
+                "Provided API key does not have WRITE permissions.".to_string(),
```
Comment (DevMattG) [inline mosaicod/crates/mosaicod-server/src/flight.rs]:
```suggestion
                "provided API key does not have WRITE permissions.".to_string(),
```

File: mosaicod/tests/src/common.rs [hunk 10]
```diff
@@ -77,20 +84,29 @@ async fn start_server(
 pub struct ServerBuilder {
     host: String,
     port: u16,
-    pool: sqlx::Pool<db::DatabaseType>,
     tls: Option<server::flight::TlsConfig>,
+    db: db::testing::Database,
+    enable_api_key: Option<bool>,
```
Comment (DevMattG) [inline mosaicod/tests/src/common.rs]: Is there a reason why you used `Option<bool>` for `enable_api_key`? Otherwise you can use directly bool (the same inside Server struct) and set it to `false` as default inside `ServerBuilder::new`

**Review by DevMattG — CHANGES_REQUESTED** [review]

File: mosaicod/crates/mosaicod-marshal/src/actions/core.rs [hunk 11]
```diff
@@ -108,6 +108,33 @@ pub enum ActionRequest {
...
+            Self::SequenceNotificationPurge(_) => write!(f, "SequenceNotificationPurge"),
+            Self::TopicCreate(_) => write!(f, "TopicCreate"),
+            Self::TopicDelete(_) => write!(f, "TopicDelete"),
+            Self::TopicNotificationCreate(_) => write!(f, "TopicNotificationCreate"),
+            Self::TopicNotificationList(_) => write!(f, "TopicNotificationList"),
+            Self::TopicNotificationPurge(_) => write!(f, "TopicNotificationPurge"),
+            Self::SessionCreate(_) => write!(f, "SessionCreate"),
+            Self::SessionFinalize(_) => write!(f, "SessionFinalize"),
+            Self::SessionDelete(_) => write!(f, "SessionDelete"),
+            Self::Query(_) => write!(f, "Query()"),
```
Comment (DevMattG) [inline mosaicod/crates/mosaicod-marshal/src/actions/core.rs]:
```suggestion
            Self::Query(_) => write!(f, "Query"),
```

**Review by DevMattG — CHANGES_REQUESTED** [review]

**Review by DevMattG — APPROVED** [review]
